### ISSUE TITLE
TS3.0 regression: error on object with similar keys set to the same required module

### ISSUE THREAD

**Issue by brendankenny** [Microsoft/TypeScript#26307]
**TypeScript Version:**  3.0.1, 3.1.0-dev.20180808
No issue in 2.9.2

**Search Terms:**
hyphen, dash, duplicate require, JSON...and I just looked at all the JS bugs since 2.9 came out :)

**Code**

*en.json*
```js
{
  "stringName": {
    "message": "message",
    "description": "message description"
  }
}
```

*locales.js*
```js
/** @type {Record<'en-US'|'en', Record<string, {message: string}>>} */
const locales = {
  'en-US': require('./en.json'),
  'en': require('./en.json'),
};

module.exports = locales;
```

Run with `tsc --strict --noEmit --allowJs --checkJs --resolveJsonModule .`

**Expected behavior:**
No errors.

**Actual behavior:**
```
error TS2322: Type '{ "stringName": { "message": string; "description": string; }; }'
  is not assignable to type 'Record<string, { message: string; }>'.
4   'en': require('./en.js'),
    ~~~~
```

I can't replicate the issue in an equivalent typescript file and the test case doesn't seem to be able to get much simpler. Deleting the `description` property in the json file or making it a local object instead of a required one makes the error go away.

With this small test case I still get the error if I rename the test file to something else (e.g. `zh.json`) or make the json file a js file with a commonjs export, but in [the larger project this comes from](https://github.com/GoogleChrome/lighthouse/blob/82ff1fedfd4ddda77583805defa4be50340c648c/lighthouse-core/lib/locales/index.js#L14) those changes make the error go away.

**Comment by brendankenny** [comment]
Unfortunately can't bisect back past #25498's crash fixed in https://github.com/Microsoft/TypeScript/commit/6b4a3b2bbb4ddacd641ba655eee28295a49e80a9, but looks like that might be related anyways :)

(before 84f5aa540e650f2e291eeb3600deabec6882f507, when the crash was introduced, there was no error)

**Comment by andy-ms** [comment]
A related error: The json file having an *extra* property can mask a problem with a (should-be) unrelated object literal having too few properties.

```ts
// @allowJs: true
// @checkJs: true
// @noEmit: true
// @strict: true
// @resolveJsonModule: true

// @Filename: /json.json
{ "a": { "s": "", "extra": 0 } }

// @Filename: /js.js
module.exports = { a: 0 };

// @Filename: /user.js
/** @typedef {{ a: { s: string } }} T */

/** @type {{ x: T, y: T }} */
const withoutJson = { x: { a: { s: "", extra: 0 } } }; // Error (good)

/** @type {{ x: T, y: T }} */
const withJson = { x: require("./json.json") }; // No error (bad)
```

Should fix #26429 first though.

**Comment by andy-ms** [comment]
Looks like this was fixed by #26817. Please test again in the next `typescript@next` release. (Doesn't work in `typescript@next` right now but works with a local build.)
